content/blog/callpackage-a-tool-for-the-lazy.md

```diff
@@ -0,0 +1,219 @@
+
+Building via `nix-build -A people` will now yield a script that prints "hello,
+people" instead.
+
+We can use the very same syntax to also overwrite the automatically discovered
+arguments like `writeShellScriptBin`, though that doesn't make sense here.
+
+Though, for a Go program that expects `buildGoModule` it is common to see some
+expression like `callPackage ./go-program.nix { buildGoModule = buildGo116Module; }`
+to enforce a certain Go compiler version.
+
+## 2. Benefit: overrides
+
+As a consequence from the parametrized builds, we can also change the value of
+the parameters after the fact, using the derivations `override` function.
+
+Consider this new `default.nix`:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> { }; in
+rec {
+  hello = pkgs.callPackage ./hello.nix { };
+  people = pkgs.callPackage ./hello.nix { audience = "people"; };
+  folks = hello.override { audience = "folks"; };
+}
+```
+
+Building and running the `folks` attribute, will give again a new version of the
+script.
+
+All the other parameters will remain the same as the have been when `hello` was
+instantiated.
+
+This is especially useful and often seen on packages that provide a whole lot of
+options to optimize the build.
+
+An example to mention here is the `neovim` attribute in nixpkgs, which has has 
+some overrideable arguments like `extraLuaPackages`, `extraPythonPackages`, or
+`withRuby`.
+
+## 3. Benefit: modifiable
+
+And now I want to introduce one of my most favorite benefits:
+
+You can actually create your own version of `callPackage`, which comes in quite
+handy when you have large sets where the attributes to be built depend on each
+other.
+
+```nix
+# In the next examples I will not implement or show the "called" 
+# files, as I think they are not necessary to understand the point I
+# want to make.
+```
+
+Consider the following initial version:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> { }; in
+rec {
+  a = pkgs.callPackage ./a.nix { };
+  b = pkgs.callPackage ./b.nix { inherit a; };
+  c = pkgs.callPackage ./c.nix { inherit b; };
+  d = pkgs.callPackage ./d.nix { };
+  e = pkgs.callPackage ./e.nix { inherit c d; };
+}
+```
+
+Here you have to remember passing required arguments that are not in nixpkgs'
+toplevel manually.
+
+This can become quite tedious quickly, especially the larger the set becomes.
+
+Therefore we can use `lib.callPackageWith` to create our own `callPackage` version.
+
+```nix
+# default.nix
+let
+  pkgs = import <nixpkgs> { };
+  callPackage = lib.callPackageWith (pkgs // packages);
+  packages = {
+    a = callPackage ./a.nix { };
+    b = callPackage ./b.nix { };
+    c = callPackage ./c.nix { };
+    d = callPackage ./d.nix { };
+    e = callPackage ./e.nix { };
+  };
+in
+  packages
+```
+
+Our modified `callPackage` now will exactly "know" how to resolve the dependencies
+through the set defined by `pkgs // packages`.
+
+Nix' laziness does us a good favour here and makes this actually possible.
```

Comment:
```suggestion

First of all note that instead of a recursive attribute set, the names we operate on are now assigned in a `let` binding.
It has the same property as recursive sets: names on the left can be used in expressions on the right of the equal sign (`=`).

This is how we can refer to `packages` when we merge its contents with the pre-existing attribute set `pkgs` using the `//` operator.

Our custom `callPackages` now makes available all the attributes in `pkgs` *and* `packages` to the called package function (the same names from `packages` taking precedence), and  `packages` is being built up recursively with each call.

The last bit may make your head spin. This construction is only possible because the Nix language is lazily evaluated. That is, values are only computed when they are actually needed. It allows passing `packages` around without having fully defined it.

Each package's dependencies are now implicit at this level (they are still explicit in each of the package files), and `callPackage` "knows" how to resolve them.
This relieves us from dealing with them manually, and precludes configuration errors that may only surface late into a lengthy build process.

Of course this small example is still manageable in the original form, and the implicitly recursive variant probably obscures the structure for software developers not familiar with lazy evaluation, making it harder to read for them than it was before.

But this benefit really pays off for large constructions, where it is the amount of code that would obscure the structure, and where manual modifications would become cumbersome and error-prone.
```


```